ambry-cloud/src/main/java/com.github.ambry.cloud/HelixVcrCluster.java

```diff
@@ -0,0 +1,130 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Helix Based VCR Cluster.
+ */
+public class HelixVcrCluster implements VirtualReplicatorCluster {
+  private final static Logger logger = LoggerFactory.getLogger(HelixVcrCluster.class);
+  private final DataNodeId currentDataNode;
+  private final String vcrClusterName;
+  private final String vcrInstanceName;
+  private final HelixManager manager;
+  private final HelixAdmin helixAdmin;
+  private final Map<String, PartitionId> partitionIdMap;
+  private final HashSet<PartitionId> assignedPartitionIds = new HashSet<>();
+  private final HelixVcrClusterMetrics metrics;
+
+  /**
+   * Construct the helix VCR cluster.
+   * @param cloudConfig The cloud configuration to use.
+   * @param clusterMapConfig The clustermap configuration to use.
+   * @param clusterMap The clustermap to use.
+   */
+  public HelixVcrCluster(CloudConfig cloudConfig, ClusterMapConfig clusterMapConfig, ClusterMap clusterMap)
+      throws Exception {
+    if (Utils.isNullOrEmpty(cloudConfig.VCR_CLUSTER_ZK_CONNECT_STRING)) {
+      throw new IllegalArgumentException("Missing value for " + CloudConfig.VCR_CLUSTER_ZK_CONNECT_STRING);
+    } else if (Utils.isNullOrEmpty(cloudConfig.VCR_CLUSTER_NAME)) {
+      throw new IllegalArgumentException("Missing value for " + CloudConfig.VCR_CLUSTER_NAME);
+    } else if (Utils.isNullOrEmpty(cloudConfig.VCR_CLUSTER_ZK_CONNECT_STRING)) {
+      throw new IllegalArgumentException("Missing value for " + CloudConfig.VCR_CLUSTER_ZK_CONNECT_STRING);
+    }
+
+    currentDataNode = new CloudDataNode(cloudConfig, clusterMapConfig);
+    List<? extends PartitionId> allPartitions = clusterMap.getAllPartitionIds(null);
+    logger.trace("All partitions from clusterMap: " + allPartitions);
+    partitionIdMap = allPartitions.stream().collect(Collectors.toMap(PartitionId::toPathString, Function.identity()));
+    vcrClusterName = cloudConfig.vcrClusterName;
+    vcrInstanceName =
+        ClusterMapUtils.getInstanceName(clusterMapConfig.clusterMapHostName, clusterMapConfig.clusterMapPort);
+
+    manager = HelixManagerFactory.getZKHelixManager(vcrClusterName, vcrInstanceName, InstanceType.PARTICIPANT,
+        cloudConfig.vcrClusterZkConnectString);
+    manager.getStateMachineEngine()
+        .registerStateModelFactory(LeaderStandbySMD.name, new HelixVcrStateModelFactory(this));
+    manager.connect();
+    helixAdmin = manager.getClusterManagmentTool();
+    metrics = new HelixVcrClusterMetrics(clusterMap.getMetricRegistry());
+    logger.info("HelixVcrCluster started successfully.");
+  }
+
+  synchronized public void addPartition(String partitionIdStr) {
+    if (partitionIdMap.containsKey(partitionIdStr)) {
+      assignedPartitionIds.add(partitionIdMap.get(partitionIdStr));
+      logger.info("Added partition {} to current VCR: ", partitionIdStr);
+    } else {
+      logger.trace("Partition {} not in clusterMap on add.", partitionIdStr);
```

Comment:
Is this an error condition or just timing?  When will it happen?

Reply:
It's an error case.

for the case that VCRCluster and mainCluster(helix/static) are out of sync.

Reply:
In that case, change trace to error or at least warn.